### PR TITLE
MDEV-20125 Redact MYSQL_PWD environment variable under Linux

### PR DESCRIPTION
**Pull request by gsauthof**
Under Linux, when using the `mysql` client binary, passing the
password via the MYSQL_PWD environment variable is more secure
than passing it on the command line with the --password option,
because:

- the argument vector is world-readable via /proc/$pid/cmdline
- although the cmdline password is overwritten by the `mysql`
  process (with a bunch of `x` characters),
  there is still a time window after process start and overwrite
  where the password can be observed by any user
- in contrast to that, the environment vector is only readable by
  the `mysql` invoking user and root

But having the password in cleartext easily accesible via
/proc/$pid/environ (or via an equivalent ps command) - even for a
restricted set of users - for the complete `mysql` client
runtime is suboptimal on sensitive systems.

As a defense-in-depth measure, this change redacts the MYSQL_PWD
environment variable under Linux. Thus, this increases the effort
for an attacker to read out the password (in contrast to just
being able to read out /proc/$pid/environ).

Note that this approach might also be effective on other systems,
but I've just tested it on Linux.

This change complements how the `mysql` client deals with
passwords supplied on the command line (cf. the --password option).

### CODE REVIEW

**Comment by gsauthof** [comment]
How to test this change:

Create DB test user - say - juser

Connect with the command line client, e.g.:

    $ MYSQL_PWD=geheim /path/to/bin/mysql --socket path/to/mysql.socket --user juser

Check (in another terminal) that the environment variable is indeed redacted, e.g.:

    $ < /proc/$(pidof mysql)/environ tr '\0' '\n' | grep MYSQL_PWD
    MYSQL_PWD=xxxxxx

**Comment by an3l** [comment]
Hi @gsauthof,
Thanks for your contribution.

Similar to other open source projects, the MariaDB Foundation needs to have shared ownership of all code that is included in the MariaDB distribution. The easiest way to achieve this is by submitting your code under the BSD-new license.

The other alternative is to sign the code contribution agreement which can be found here: https://mariadb.com/kb/en/mariadb/mca/

Please indicate in a comment below, or update your PR comment that you are contributing your new code of the whole pull request, including one or several files that are either new files or modified ones, under the BSD-new license or that you have filled out the contribution agreement and sent it.
Thanks,
Anel

**Comment by gsauthof** [comment]
I contribute my code in this pull request under the BSD-new license.

File: client/mysql.cc [hunk 1]
```diff
@@ -1116,6 +1116,24 @@ inline int get_command_index(char cmd_char)
...
+  char *passwd = getenv("MYSQL_PWD");
+  if (passwd) {
+      /* setenv copies passwd, result not visible in /proc/$pid/environ */
+      if (setenv("MYSQL_PWD", passwd, 1) == -1) {
+        put_info("setenv() failed", INFO_ERROR, 0);
+        my_end(0);
+        exit(1);
+      }
+      /* overwrite password in /proc/$pid/environ */
+      memset(passwd, 'x', strlen(passwd));
```
Comment (grooverdan) [inline client/mysql.cc]: is there a way to overwrite this without exposing the length of the password? Does this work on Windows/BSD?

File: client/mysql.cc [hunk 2]
```diff
@@ -1116,6 +1116,24 @@ inline int get_command_index(char cmd_char)
...
 }
 
+static void redact_mysql_pwd(void)
+{
+#ifndef DONT_USE_MYSQL_PWD
+  /* defense-in-depth: overwrite password in original environment vector */
+  char *passwd = getenv("MYSQL_PWD");
+  if (passwd) {
+      /* setenv copies passwd, result not visible in /proc/$pid/environ */
+      if (setenv("MYSQL_PWD", passwd, 1) == -1) {
```
Comment (grooverdan) [inline client/mysql.cc]: By overwriting the password here you've effectively disabled its use entirely (in mysql_real_connect in libmysqld/libmysqld.c).

File: client/mysql.cc [hunk 3]
```diff
@@ -1128,6 +1146,8 @@ int main(int argc,char *argv[])
   MY_INIT(argv[0]);
   DBUG_ENTER("main");
   DBUG_PROCESS(argv[0]);
+
+  redact_mysql_pwd();
```
Comment (grooverdan) [inline client/mysql.cc]: so early in the process effectively disables its use.

**Review by grooverdan — CHANGES_REQUESTED** [review]
An implementation in libmysqld/libmysqld.c would seem like a better location to hides its use in other tools like mysqldump etc. Need to carefully look at all locations to ensure that if a second connection occurs that existing behaviour that has `MYSQL_PWD` working continues to work.

You might need to add a `mysql->option` to clear/preserve `MYSQL_PWD` to preserve the behaviour of implementations that do reconnect.

File: client/mysql.cc (hunk 3 above)
Comment (grooverdan) [reply client/mysql.cc]: note: used by scripts/wsrep_sst_mysqldump.sh

File: client/mysql.cc (hunk 1 above)
Comment (gsauthof) [reply client/mysql.cc]: I don't know a method where the length isn't exposed. FWIW, when passing the password in the argument vector (with -p/--password) the length is exposed, as well. I don't know if this is also effective on Windows/BSD. I don't even know if those systems expose the environments variables to other processes in a similar way.

File: client/mysql.cc (hunk 2 above)
Comment (gsauthof) [reply client/mysql.cc]: I don't understand your comment. I overwrite the password with itself such that any following `getenv("MYSQL_PWD")` still retrieves the original value. It's just overwritten in the original environment vector.

File: client/mysql.cc (hunk 3 above)
Comment (gsauthof) [reply client/mysql.cc]: See my previous comment, a following `getenv("MYSQL_PWD")` still returns the original password.

**Comment by cvicentiu** [comment]
Hi @gsauthof!

I have looked at this pull request in depth, tested it both on Linux and Windows. While this change does not cause immediate problems, I have problems accepting it.

All documentation of getenv, be it for windows, linux or OS X say the caller should not modify the memory at the address returned by getenv.

Although the effect of the PR is indeed that the /proc/<pid>/environ values are alterered, I feat that this is a ticking bomb as we are misusing the API. Furthermore, (not on UNIX systems in particular, but generally, the C standard) warns that modifying the string is undefined behaviour.
See:
https://en.cppreference.com/w/c/program/getenv

With this said, I understand the security aspects this patch is trying to improve, but since proc environ is only readable by the user that started the process, or root, adding to the fact that there are significant alternatives to a password-less login such as unix_socket plugin or the pam plugin I feel it is not wise for us to delve into maintaining an undefined behaviour change that may end up causing other problems down the line.

If you have other suggestions how to improve this, without causing undefined behaviour, feel free to submit a pull request. For now I will close the pull request.

Thank you for the time spent working on this!

**Comment by ericherman** [comment]
Unfortunate, but I believe @cvicentiu made the correct call.

If undefined behaviour is required and also if this undefined behaviour truly can be relied upon, then I'd like to see a very strong case made.

**Comment by gsauthof** [comment]
Hey @cvicentiu and @ericherman  - an alternative way to remove the sensitive password from `/proc/$pid/environ` is to use the official Linux API.

I've created a small example that demonstrates this approach:

https://gist.github.com/2691e13bbaec23af6a3931ceb139c325

Would a solution based on this approach be something you would accept in MariaDB?

Advantages of this approach:

- doesn't rely on system behavior that is specified as undefined behavior in the C standard (or unspecified in POSIX)
- uses official Linux API that comes with strong compatibility guarantees
- as portable as the previous approach (the previous approach likely only is effective/relevant on Linux, anyways)

Disadvantages:

- requires slightly more code

**Comment by cvicentiu** [comment]
Hi @gsauthof!

I need a bit of time to digest this. I only just saw your message now. I'll try to give you an answer as soon as possible.

**Pull request closed without merging**